### ISSUE TITLE
[Question] Load Generic Types

### ISSUE THREAD

**Issue by tommelo** [classgraph/classgraph#244]
Hey folks,

I'm using **classgraph** in a project and I just got stuck in a scenario with generic types.

Here's the scenario:

```
@MyAnnotation
public class MyClass implements MyInterface<Type1, Type2> {

}
```

Steps to perform:

1 - Find all classes annotated with @MyAnnotation;
2 - Check if the class implements the interface MyInterface;
3 - Get the parameterized input and output(Type1 and Type2) classes.

I was able to accomplish this with the following code:

```
ScanResult scanResult = new ClassGraph()
    .enableAllInfo()
    .whitelistPackages("my.package")
    .scan();

String annotationClassName = MyAnnotation.class.getCanonicalName();
ClassInfoList routes = scanResult.getClassesWithAnnotation(annotationClassName);

for (ClassInfo info : routes) {
    ClassTypeSignature typeSignature = info.getTypeSignature();

    for (ClassRefTypeSignature ref : typeSignature.getSuperinterfaceSignatures()) {

        if (MyInterface.class.getName().equals(ref.getBaseClassName())) {
            List<TypeArgument> args = ref.getTypeArguments();
            Class<?> inputClass = Class.forName(args.get(0).toString());
            Class<?> outputClass = Class.forName(args.get(1).toString());
        } 

    }
}
```
This works fine for non-generic types, but if I have an implementation like:
```
@MyAnnotation
public class MyClass implements MyInterface<Type1, List<Type2>> {

}
```
This code won't work because the object doesn't know about its generic type at execution time, so when I try Class.forName(args.get(1).toString()) I'll get ClassNotFoundException.

So my question is: Is there a better approach using classgraph to accomplish these 3 steps? Does classgraph offers a **loadClass()** method for TypeArgument?

**Comment by lukehutch** [comment]
@tommelo I'm always happy to see someone actually using the type signature APIs in ClassGraph, because they were a ton of work but some of the most fun bits of ClassGraph to write -- but almost nobody uses them :-)

For a `TypeArgument`, you have two methods: `getWildcard()` (returning `NONE`, `ANY` (same as `?`), `EXTENDS` and `SUPER`), and then `getTypeSignature()`, which returns a `ReferenceTypeSignature`. `ReferenceTypeSignature` has the following concrete subclasses:

* `ArrayTypeSignature` -- depending on whether or not your interface even allows array-typed parameterizations, you may need to handle array parameter types specially, e.g. `MyInterface<String, int[]>`, but you can get the array element type and number of dimensions from the `ArrayTypeSignature` methods;
* `TypeVariableSignature` -- which actually has a handy `resolve()` method that will resolve the type variable against the type parameters of the containing method, and if that fails, against the type parameters of the containing class, returning another `TypeParameter` that you can recursively process the same way as the above;
* `ClassRefTypeSignature` -- which has the `loadClass()` method you're looking for.

So basically, you had 95% of it figured out (qudos for that) -- you were only one or two steps away from what you needed.

Please let me know if this does what you need.

**Comment by tommelo** [comment]
@lukehutch wow, that's a ton of work, indeed! Great job!

So close! I actually tried ```getWildcard()``` at first, but I got ```NONE``` so I thought "well, this may not be the way to go" =)

I'll try ```getWildcard()``` and ```getTypeSignature()``` as well, probably this will get things done! I'll get back to you as soon as possible.

Thank you!
I really appreciate your time for this detailed explanation!

**Comment by tommelo** [comment]
@lukehutch ```getTypeSignature()``` worked like a charm!

Thanks for pointing out the right direction!
**ClassGraph** is really great!

We can close this issue.

**Comment by tommelo** [comment]
@lukehutch that saved me some lines of code! Thanks again!

So far I could get a few tests working with: ```((ClassRefTypeSignature) args.get(0).getTypeSignature()).loadClass();```, Now I'm able to serialize/deserialize types like: ```List<MyClass>```. I'm pretty sure that I'll get more complex cases like ```MyInterface<String, ? extends Parameter>``` but a few limitations it's ok for the first release.

I'll keep posting in this issue as I go through some complex scenarios.
Let me know if you need any help!!

**Comment by lukehutch** [comment]
There's a bigger problem at hand, which is that the type parameters of supertypes and interfaces are not resolved at present. For example, for `interface A<X, Y>` and `class B<Z> implements A<Integer, Z>`, the interface implemented by `B` is actually `A<Integer, Z>`, not `A<X, Y>`. And if `A` is implemented by multiple classes with different type arguments, it can take different concrete type parameter values, so the concrete types cannot be stored in the `ClassInfo` for `A` (including in the type signature stored in the `ClassInfo` for `A`). There has to be a way to get the concrete resolved type signatures of all interfaces and superclasses of a given base class (`B` in this example).

This also extends to methods and fields too -- if you have a method in `A` of `public X getX()`, then that should get resolved to `public Integer getX()` in this example, using `B` as the base class for type resolution.

(In fact another thing I realized is that the `ClassInfo#getMethodInfo()` and `ClassInfo#getFieldInfo()` methods right now only return declared methods, not methods inherited from superclasses (or declared in interfaces, in the case of abstract methods).

So this is taking quite a bit of work to implement and get right, but I'm glad you filed this bug report, since it made me realize the API is incomplete! (And actually, for your purposes, you will get wrong type parameter values in cases like the above, e.g. you will get a type variable when you expected a concrete type.)

**Comment by johnou** [comment]
Sounds like a performance killer collecting all that information?

**Comment by lukehutch** [comment]
The old API will remain. If you need properly resolved types, you can use the new calls. They will be a bit slower than the old calls, but you can't put a price on correctness. And the increase in time to compute the right type info will be nothing compared to the I/O time already needed to read the classfiles in the first place.

**Comment by lukehutch** [comment]
Turns out this is extremely complex. If you have `class X<Y, Z extends List & Y> extends S<Z>`, then the fields of type `Z` in class `S` have the type bounds of `extends List implements Y`, and there's no way to declare this type in Java source (you can only write code where the field conforms to one of `List` or `Y`).

So actually I'm going to punt the resolution of generic types to the user, if they care enough to resolve type arguments using type parameters.

I'll also make a note in the Javadoc for `ClassInfo#getMethodInfo()` and `ClassInfo#getFieldInfo()` that only declared methods and fields are returned -- you need to iterate to supertypes to get all methods/fields.
